🐛 [BUG] Compatibility with torch1.9, geometric1.7, e3nn 0.3.2

**Describe the bug**
torch.jit.script cannot take dict for typing.Dict .

**To Reproduce**
Unit test tests/data/test_AtomicData.py::test_non_periodic_edge  will fail with pytorch1.9 and torch_geometric 1.7

**Environment (please complete the following information):**
 - OS: Fedora
 - python version 3.8.8
 - python environment:
   - nequip version 0.3.2
   - e3nn version 0.3.2
   - pytorch version 1.9.0+cu102

Pytest error message:

`========================================== FAILURES ==========================================
______________________________ test_non_periodic_edge[float32] _______________________________

CH3CHO = (Atoms(symbols='OCHCH3', pbc=False), AtomicData(edge_index=[2, 18], pos=[7, 3], num_nodes=7, atomic_numbers=[7], cell=[3, 3], edge_cell_shift=[18, 3], pbc=[3]))

    def test_non_periodic_edge(CH3CHO):
        atoms, data = CH3CHO
        # check edges
        for edge in range(data.num_edges):
            real_displacement = (
                atoms.positions[data.edge_index[1, edge]]
                - atoms.positions[data.edge_index[0, edge]]
            )
>           assert torch.allclose(
                data.get_edge_vectors()[edge],
                torch.as_tensor(real_displacement, dtype=torch.get_default_dtype()),
            )

tests/data/test_AtomicData.py:30: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

data = AtomicData(edge_index=[2, 18], pos=[7, 3], num_nodes=7, atomic_numbers=[7], cell=[3, 3], edge_cell_shift=[18, 3], pbc=[3])

    def get_edge_vectors(data: Data) -> torch.Tensor:
>       data = AtomicDataDict.with_edge_vectors(AtomicData.to_AtomicDataDict(data))
E       RuntimeError: with_edge_vectors() Expected a value of type 'Dict[str, Tensor]' for argument 'data' but instead found type 'dict'.
E       Position: 0
E       Value: {'atomic_numbers': tensor([8, 6, 1, 6, 1, 1, 1]), 'num_nodes': 7, 'pos': tensor([[ 1.2181,  0.3612,  0.0000],
E               [ 0.0000,  0.4641,  0.0000],
E               [-0.4772,  1.4653,  0.0000],
E               [-0.9481, -0.7001,  0.0000],
E               [-0.3859, -1.6342,  0.0000],
E               [-1.5963, -0.6525,  0.8809],
E               [-1.5963, -0.6525, -0.8809]]), 'edge_index': tensor([[0, 1, 1, 1, 2, 3, 3, 3, 3, 4, 4, 4, 5, 5, 5, 6, 6, 6],
E               [1, 3, 0, 2, 1, 5, 6, 4, 1, 3, 5, 6, 6, 4, 3, 5, 3, 4]]), 'edge_cell_shift': tensor([[0., 0., 0.],
E               [0., 0., 0.],
E               [0., 0., 0.],
E               [0., 0., 0.],
E               [0., 0., 0.],
E               [0., 0., 0.],
E               [0., 0., 0.],
E               [0., 0., 0.],
E               [0., 0., 0.],
E               [0., 0., 0.],
E               [0., 0., 0.],
E               [0., 0., 0.],
E               [0., 0., 0.],
E               [0., 0., 0.],
E               [0., 0., 0.],
E               [0., 0., 0.],
E               [0., 0., 0.],
E               [0., 0., 0.]]), 'pbc': tensor([False, False, False]), 'cell': tensor([[0., 0., 0.],
E               [0., 0., 0.],
E               [0., 0., 0.]])}
E       Declaration: with_edge_vectors(Dict(str, Tensor) data, bool with_lengths=True) -> (Dict(str, Tensor))
E       Cast error details: Unable to cast Python instance to C++ type (compile in debug mode for details)

nequip/data/AtomicData.py:253: RuntimeError
`

without pull #46 ,  with the unit test in the development branch will fail at tests/data/test_AtomicData.py::test_non_periodic_edge with error message:

```
tests/data/test_AtomicData.py::test_non_periodic_edge[float32]

====================================================== FAILURES ======================================================
__________________________________________ test_non_periodic_edge[float32] ___________________________________________

CH3CHO = (Atoms(symbols='OCHCH3', pbc=False), AtomicData(edge_index=[2, 18], pos=[7, 3], num_nodes=7, atomic_numbers=[7], cell=[3, 3], edge_cell_shift=[18, 3], pbc=[3]))

    def test_non_periodic_edge(CH3CHO):
        atoms, data = CH3CHO
        # check edges
        for edge in range(data.num_edges):
            real_displacement = (
                atoms.positions[data.edge_index[1, edge]]
                - atoms.positions[data.edge_index[0, edge]]
            )
>           assert torch.allclose(
                data.get_edge_vectors()[edge],
                torch.as_tensor(real_displacement, dtype=torch.get_default_dtype()),
            )

tests/data/test_AtomicData.py:30: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

data = AtomicData(edge_index=[2, 18], pos=[7, 3], num_nodes=7, atomic_numbers=[7], cell=[3, 3], edge_cell_shift=[18, 3], pbc=[3])

    def get_edge_vectors(data: Data) -> torch.Tensor:
>       data = AtomicDataDict.with_edge_vectors(AtomicData.to_AtomicDataDict(data))
E       RuntimeError: with_edge_vectors() Expected a value of type 'Dict[str, Tensor]' for argument 'data' but instead found type 'dict'.
E       Position: 0
E       Value: {'cell': tensor([[0., 0., 0.],
E               [0., 0., 0.],
E               [0., 0., 0.]]), 'pbc': tensor([False, False, False]), 'atomic_numbers': tensor([8, 6, 1, 6, 1, 1, 1]), 'num_nodes': 7, 'pos': tensor([[ 1.2181,  0.3612,  0.0000],
E               [ 0.0000,  0.4641,  0.0000],
E               [-0.4772,  1.4653,  0.0000],
E               [-0.9481, -0.7001,  0.0000],
E               [-0.3859, -1.6342,  0.0000],
E               [-1.5963, -0.6525,  0.8809],
E               [-1.5963, -0.6525, -0.8809]]), 'edge_cell_shift': tensor([[0., 0., 0.],
E               [0., 0., 0.],
E               [0., 0., 0.],
E               [0., 0., 0.],
E               [0., 0., 0.],
E               [0., 0., 0.],
E               [0., 0., 0.],
E               [0., 0., 0.],
E               [0., 0., 0.],
E               [0., 0., 0.],
E               [0., 0., 0.],
E               [0., 0., 0.],
E               [0., 0., 0.],
E               [0., 0., 0.],
E               [0., 0., 0.],
E               [0., 0., 0.],
E               [0., 0., 0.],
E               [0., 0., 0.]]), 'edge_index': tensor([[0, 1, 1, 1, 2, 3, 3, 3, 3, 4, 4, 4, 5, 5, 5, 6, 6, 6],
E               [1, 3, 0, 2, 1, 5, 6, 4, 1, 3, 5, 6, 6, 4, 3, 5, 3, 4]])}
E       Declaration: with_edge_vectors(Dict(str, Tensor) data, bool with_lengths=True) -> (Dict(str, Tensor))
E       Cast error details: Unable to cast Python instance to C++ type (compile in debug mode for details)

nequip/data/AtomicData.py:253: RuntimeError
```

following the previous set up, another error will show up at tests/data/test_AtomicData.py::test_without_nodes

```
====================================================== FAILURES ======================================================
____________________________________________ test_without_nodes[float32] _____________________________________________

CH3CHO = (Atoms(symbols='OCHCH3', pbc=False), AtomicData(edge_index=[2, 18], pos=[7, 3], num_nodes=7, atomic_numbers=[7], cell=[3, 3], edge_cell_shift=[18, 3], pbc=[3]))

    def test_without_nodes(CH3CHO):
        # Non-periodic
        atoms, data = CH3CHO
        which_nodes = [0, 5, 6]
>       new_data = data.without_nodes(which_nodes=which_nodes)

tests/data/test_AtomicData.py:59: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = AtomicData(edge_index=[2, 18], pos=[7, 3], num_nodes=7, atomic_numbers=[7], cell=[3, 3], edge_cell_shift=[18, 3], pbc=[3])
which_nodes = tensor([0, 5, 6])

    def without_nodes(self, which_nodes):
        """Return a copy of ``self`` with ``which_nodes`` removed.
        The returned object may share references to some underlying data tensors with ``self``.
        Args:
            which_nodes (index tensor or boolean mask)
        Returns:
            A new data object.
        """
        which_nodes = torch.as_tensor(which_nodes)
        if which_nodes.dtype == torch.bool:
            mask = ~which_nodes
        else:
            mask = torch.ones(self.num_nodes, dtype=torch.bool)
            mask[which_nodes] = False
        assert mask.shape == (self.num_nodes,)
        n_keeping = mask.sum()
    
        # Only keep edges where both from and to are kept
        edge_mask = mask[self.edge_index[0]] & mask[self.edge_index[1]]
        # Create an index mapping:
        new_index = torch.full((self.num_nodes,), -1, dtype=torch.long)
        new_index[mask] = torch.arange(n_keeping, dtype=torch.long)
    
        new_dict = {}
        for k in self.keys:
            if k == AtomicDataDict.EDGE_INDEX_KEY:
                new_dict[AtomicDataDict.EDGE_INDEX_KEY] = new_index[
                    self.edge_index[:, edge_mask]
                ]
            elif k == AtomicDataDict.EDGE_CELL_SHIFT_KEY:
                new_dict[AtomicDataDict.EDGE_CELL_SHIFT_KEY] = self.edge_cell_shift[
                    edge_mask
                ]
            elif k == AtomicDataDict.CELL_KEY:
                new_dict[k] = self[k]
            else:
>               if len(self[k]) == self.num_nodes:
E               TypeError: object of type 'int' has no len()

nequip/data/AtomicData.py:325: TypeError
```

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -7,6 +7,7 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 Most recent change on the bottom.
 
 ## [Unreleased]
+- update interfaces with torch_geometric 1.7 and e3nn 0.3.2
 
 ## [0.3.2] - 2021-06-09
 ### Added

--- a/README.md
+++ b/README.md
@@ -29,7 +29,7 @@ pip install e3nn==0.3.2
 
 where ```${CUDA}``` should be replaced by either ```cpu```, ```cu101```, ```cu102```, or ```cu111``` depending on your PyTorch installation, for details see [here](https://github.com/rusty1s/pytorch_geometric). 
 
-* Install [e3nn](https://github.com/e3nn/e3nn), version 0.2.9: 
+* Install [e3nn](https://github.com/e3nn/e3nn), version 0.3.2: 
 
 ```
 pip install e3nn==0.3.2

--- a/README.md
+++ b/README.md
@@ -24,15 +24,15 @@ pip install torch-sparse -f https://pytorch-geometric.com/whl/torch-1.8.0+${CUDA
 pip install torch-cluster -f https://pytorch-geometric.com/whl/torch-1.8.0+${CUDA}.html
 pip install torch-spline-conv -f https://pytorch-geometric.com/whl/torch-1.8.0+${CUDA}.html
 pip install torch-geometric==1.7.1
-pip install e3nn==0.3.2
+pip install e3nn==0.3.3
 ```
 
 where ```${CUDA}``` should be replaced by either ```cpu```, ```cu101```, ```cu102```, or ```cu111``` depending on your PyTorch installation, for details see [here](https://github.com/rusty1s/pytorch_geometric). 
 
-* Install [e3nn](https://github.com/e3nn/e3nn), version 0.3.2: 
+* Install [e3nn](https://github.com/e3nn/e3nn), version 0.3.3:
 
 ```
-pip install e3nn==0.3.2
+pip install e3nn==0.3.3
 ```
 
 * Install our fork of [`pytorch_ema`](https://github.com/Linux-cpp-lisp/pytorch_ema) for using an Exponential Moving Average on the weights: 

--- a/README.md
+++ b/README.md
@@ -16,15 +16,15 @@ NequIP requires:
 
 To install:
 
-* Install [PyTorch Geometric](https://github.com/rusty1s/pytorch_geometric), make sure to install this with your correct version of CUDA/CPU and to use PyTorch Geometric version 1.7.0:
+* Install [PyTorch Geometric](https://github.com/rusty1s/pytorch_geometric), make sure to install this with your correct version of CUDA/CPU and to use PyTorch Geometric version 1.7.1:
 
 ```
 pip install torch-scatter -f https://pytorch-geometric.com/whl/torch-1.8.0+${CUDA}.html
 pip install torch-sparse -f https://pytorch-geometric.com/whl/torch-1.8.0+${CUDA}.html
 pip install torch-cluster -f https://pytorch-geometric.com/whl/torch-1.8.0+${CUDA}.html
 pip install torch-spline-conv -f https://pytorch-geometric.com/whl/torch-1.8.0+${CUDA}.html
-pip install torch-geometric==1.7.0
-pip install e3nn==0.2.9
+pip install torch-geometric==1.7.1
+pip install e3nn==0.3.2
 ```
 
 where ```${CUDA}``` should be replaced by either ```cpu```, ```cu101```, ```cu102```, or ```cu111``` depending on your PyTorch installation, for details see [here](https://github.com/rusty1s/pytorch_geometric). 
@@ -32,7 +32,7 @@ where ```${CUDA}``` should be replaced by either ```cpu```, ```cu101```, ```cu10
 * Install [e3nn](https://github.com/e3nn/e3nn), version 0.2.9: 
 
 ```
-pip install e3nn==0.2.9
+pip install e3nn==0.3.2
 ```
 
 * Install our fork of [`pytorch_ema`](https://github.com/Linux-cpp-lisp/pytorch_ema) for using an Exponential Moving Average on the weights: 

--- a/nequip/data/AtomicData.py
+++ b/nequip/data/AtomicData.py
@@ -328,10 +328,8 @@ class AtomicData(Data):
                 ]
             elif k == AtomicDataDict.CELL_KEY:
                 new_dict[k] = self[k]
-            elif k == "num_nodes":
-                pass
             else:
-                if len(self[k]) == self.num_nodes:
+                if isinstance(self[k], torch.Tensor) and len(self[k]) == self.num_nodes:
                     new_dict[k] = self[k][mask]
                 else:
                     new_dict[k] = self[k]

--- a/nequip/data/AtomicData.py
+++ b/nequip/data/AtomicData.py
@@ -263,8 +263,15 @@ class AtomicData(Data):
             keys = data.keys()
         else:
             raise ValueError(f"Invalid data `{repr(data)}`")
+
         return {
-            k: data[k] for k in keys if (k not in exclude_keys and data[k] is not None)
+            k: data[k]
+            for k in keys
+            if (
+                k not in exclude_keys
+                and data[k] is not None
+                and isinstance(data[k], torch.Tensor)
+            )
         }
 
     @classmethod
@@ -321,6 +328,8 @@ class AtomicData(Data):
                 ]
             elif k == AtomicDataDict.CELL_KEY:
                 new_dict[k] = self[k]
+            elif k == "num_nodes":
+                pass
             else:
                 if len(self[k]) == self.num_nodes:
                     new_dict[k] = self[k][mask]

--- a/nequip/data/dataset.py
+++ b/nequip/data/dataset.py
@@ -295,8 +295,8 @@ class AtomicInMemoryDataset(AtomicDataset):
         unbiased: bool = True,
         modes: Optional[List[Union[str]]] = None,
     ) -> List[tuple]:
-        if self.__indices__ is not None:
-            selector = torch.as_tensor(self.__indices__)[::stride]
+        if self._indices is not None:
+            selector = torch.as_tensor(self._indices)[::stride]
         else:
             selector = torch.arange(0, self.len(), stride)
 
@@ -318,7 +318,7 @@ class AtomicInMemoryDataset(AtomicDataset):
                 obj = self.fixed_fields
             else:
                 obj = self.data
-
+            
             if callable(field):
                 arr = field(obj)
             else:

--- a/nequip/data/dataset.py
+++ b/nequip/data/dataset.py
@@ -318,7 +318,7 @@ class AtomicInMemoryDataset(AtomicDataset):
                 obj = self.fixed_fields
             else:
                 obj = self.data
-            
+
             if callable(field):
                 arr = field(obj)
             else:

--- a/nequip/datasets/aspirin.py
+++ b/nequip/datasets/aspirin.py
@@ -32,6 +32,6 @@ class AspirinDataset(AtomicInMemoryDataset):
         fixed_fields = {
             AtomicDataDict.ATOMIC_NUMBERS_KEY: np.asarray(data["z"], dtype=np.int),
             AtomicDataDict.PBC_KEY: np.array([False, False, False]),
-            AtomicDataDict.CELL_KEY: None,
+        #    AtomicDataDict.CELL_KEY: None,
         }
         return arrays, fixed_fields

--- a/nequip/datasets/aspirin.py
+++ b/nequip/datasets/aspirin.py
@@ -32,6 +32,6 @@ class AspirinDataset(AtomicInMemoryDataset):
         fixed_fields = {
             AtomicDataDict.ATOMIC_NUMBERS_KEY: np.asarray(data["z"], dtype=np.int),
             AtomicDataDict.PBC_KEY: np.array([False, False, False]),
-        #    AtomicDataDict.CELL_KEY: None,
+            #    AtomicDataDict.CELL_KEY: None,
         }
         return arrays, fixed_fields

--- a/nequip/datasets/aspirin.py
+++ b/nequip/datasets/aspirin.py
@@ -2,9 +2,6 @@ import numpy as np
 
 from os.path import dirname, basename, abspath
 
-from ase import units
-from ase.io import read
-
 from nequip.data import AtomicDataDict, AtomicInMemoryDataset
 
 
@@ -32,6 +29,5 @@ class AspirinDataset(AtomicInMemoryDataset):
         fixed_fields = {
             AtomicDataDict.ATOMIC_NUMBERS_KEY: np.asarray(data["z"], dtype=np.int),
             AtomicDataDict.PBC_KEY: np.array([False, False, False]),
-            #    AtomicDataDict.CELL_KEY: None,
         }
         return arrays, fixed_fields

--- a/nequip/train/trainer.py
+++ b/nequip/train/trainer.py
@@ -316,7 +316,7 @@ class Trainer:
             d = self.as_dict()
             for key in list(d.keys()):
                 if not isinstance(d[key], (float, int, str, list, tuple)):
-                    d[key] = type(d[key])
+                    d[key] = repr(d[key])
             self.log_dictionary(d, name="Initialization")
 
         logging.debug("! Done Initialize Trainer")

--- a/nequip/train/trainer.py
+++ b/nequip/train/trainer.py
@@ -312,8 +312,6 @@ class Trainer:
         # initialize the optimizer and scheduler, the params will be updated in the function
         self.init()
 
-        self.statistics = {}
-
         if not (restart and append):
             d = self.as_dict()
             for key in list(d.keys()):

--- a/nequip/utils/initialization.py
+++ b/nequip/utils/initialization.py
@@ -15,8 +15,9 @@ def unit_uniform_init_(t: torch.Tensor):
 def uniform_initialize_fcs(mod: torch.nn.Module):
     """Initialize ``e3nn.nn.FullyConnectedNet``s with ``unit_uniform_init_``"""
     if isinstance(mod, e3nn.nn.FullyConnectedNet):
-        for w in mod.weights:
-            unit_uniform_init_(w)
+        for ilayer, layer in mod._modules.items():
+            for w in layer.weight:
+                unit_uniform_init_(w)
     # no need to do torch.nn.Linear, which is uniform by default
 
 

--- a/nequip/utils/initialization.py
+++ b/nequip/utils/initialization.py
@@ -15,9 +15,8 @@ def unit_uniform_init_(t: torch.Tensor):
 def uniform_initialize_fcs(mod: torch.nn.Module):
     """Initialize ``e3nn.nn.FullyConnectedNet``s with ``unit_uniform_init_``"""
     if isinstance(mod, e3nn.nn.FullyConnectedNet):
-        for ilayer, layer in mod._modules.items():
-            for w in layer.weight:
-                unit_uniform_init_(w)
+        for layer in mod:
+            unit_uniform_init_(layer.weight)
     # no need to do torch.nn.Linear, which is uniform by default
 
 
@@ -37,11 +36,13 @@ def uniform_initialize_tps(mod: torch.nn.Module):
 def xavier_initialize_fcs(mod: torch.nn.Module):
     """Initialize ``e3nn.nn.FullyConnectedNet``s and ``torch.nn.Linear``s with Xavier uniform initialization"""
     if isinstance(mod, e3nn.nn.FullyConnectedNet):
-        for w in mod.weights:
+        for layer in mod:
             # in FC:
             # h_in, _h_out = W.shape
             # W = W / h_in**0.5
-            torch.nn.init.xavier_uniform_(w, gain=w.shape[0] ** 0.5)
+            torch.nn.init.xavier_uniform_(
+                layer.weight, gain=layer.weight.shape[0] ** 0.5
+            )
     elif isinstance(mod, torch.nn.Linear):
         torch.nn.init.xavier_uniform_(mod.weight)
 
@@ -63,8 +64,8 @@ def orthogonal_initialize_linears(mod: torch.nn.Module):
 def orthogonal_initialize_fcs(mod: torch.nn.Module):
     """Initialize ``e3nn.nn.FullyConnectedNet``s and ``torch.nn.Linear``s with orthogonal initialization"""
     if isinstance(mod, e3nn.nn.FullyConnectedNet):
-        for w in mod.weights:
-            torch.nn.init.orthogonal_(w)
+        for layer in mod:
+            torch.nn.init.orthogonal_(layer.weight)
     elif isinstance(mod, torch.nn.Linear):
         torch.nn.init.orthogonal_(mod.weight)
 
@@ -72,5 +73,5 @@ def orthogonal_initialize_fcs(mod: torch.nn.Module):
 def unit_orthogonal_initialize_e3nn_fcs(mod: torch.nn.Module):
     """Initialize only ``e3nn.nn.FullyConnectedNet``s with ``unit_orthogonal_init_``"""
     if isinstance(mod, e3nn.nn.FullyConnectedNet):
-        for w in mod.weights:
-            unit_orthogonal_init_(w)
+        for layer in mod:
+            unit_orthogonal_init_(layer.weight)

--- a/setup.py
+++ b/setup.py
@@ -29,8 +29,8 @@ setup(
         "numpy",
         "ase",
         "torch>=1.8",
-        "torch_geometric",
-        "e3nn>=0.3",
+        "torch_geometric>=1.7",
+        "e3nn>=0.3.2",
         "pyyaml",
         "contextlib2;python_version<'3.7'",  # backport of nullcontext
         "typing_extensions;python_version<'3.8'",

--- a/setup.py
+++ b/setup.py
@@ -30,7 +30,7 @@ setup(
         "ase",
         "torch>=1.8",
         "torch_geometric>=1.7.1",
-        "e3nn>=0.3.2",
+        "e3nn>=0.3.3",
         "pyyaml",
         "contextlib2;python_version<'3.7'",  # backport of nullcontext
         "typing_extensions;python_version<'3.8'",

--- a/setup.py
+++ b/setup.py
@@ -29,7 +29,7 @@ setup(
         "numpy",
         "ase",
         "torch>=1.8",
-        "torch_geometric>=1.7",
+        "torch_geometric>=1.7.1",
         "e3nn>=0.3.2",
         "pyyaml",
         "contextlib2;python_version<'3.7'",  # backport of nullcontext

